spyderlib/otherplugins.py

```diff
@@ -8,52 +8,115 @@
+    loads its submodules."""
+    namespace_path = Path(plugin_path) / base_namespace / plugins_namespace
+
+    if not namespace_path.exists():
+        return
+
+    for dirname in namespace_path.dirs():
+        if dirname.name == "__pycache__":
+            continue
+        _import_plugin(dirname.name, base_namespace, plugins_namespace,
+                       namespace_path, modnames, modlist)
+
+    for name in namespace_path.files(pattern="*.py"):
+        if name.name == "__init__.py":
+            continue
+        _import_plugin(name.namebase, base_namespace, plugins_namespace,
+                       namespace_path, modnames, modlist)
+
+
+class _ModuleMock():
+    """ """
+
+
+def _import_plugin(name, base_namespace, plugin_namespace, namespace_path,
+                   modnames, modlist):
+    """Import the plugin `plugins_namsepace`.`name`, add it to `modlist` and
+    adds its name to `modname`."""
+    module_name = "{0}.{1}.{2}".format(base_namespace, plugin_namespace, name)
+
+    if module_name in modnames:
+        return
+    try:
+        # First add a mock module with the LOCALEPATH attribute so that the
+        # helper method can fin the locale on import
+        mock = _ModuleMock()
+        mock.LOCALEPATH = Path(namespace_path) / name / 'locale'
+        sys.modules[module_name] = mock
+
+        if PY33:
+            loader = importlib.machinery.PathFinder.find_module(
+                name, [namespace_path])
+            module = loader.load_module(name)
+        elif PY3:
+            spec = importlib.machinery.PathFinder.find_spec(name,
+                                                            [namespace_path])
+            module = spec.loader.load_module(name)
+        else:
+            info = imp.find_module(name, [namespace_path])
+            module = imp.load_module(module_name, *info)
+
+        # Then restore the actual loaded module instead of the mock
+        sys.modules[module_name] = module
+
+        modlist.append(module)
+        modnames.append(module_name)
+    except Exception:
+        sys.stderr.write("ERROR: 3rd party plugin import failed for "
+                         "`{0}`\n".format(module_name))
+        traceback.print_exc(file=sys.stderr)
+
+
+def get_spyderplugins_mods(io=False):
+    """Import modules from plugins package and return the list"""
+    base_namespace = "spyplugins"
+
+    if io:
+        plugins_namespace = "io"
+    else:
+        plugins_namespace = "ui"
+
+    namespace = '.'.join([base_namespace, plugins_namespace])
+
+    # Import parent module
+    importlib.import_module(namespace)
+
+    # Create user directory
+    user_conf_path = Path(get_conf_path())
+    user_plugin_basepath = user_conf_path / base_namespace
+    user_plugin_path = user_conf_path / base_namespace / plugins_namespace
+    user_plugin_path.makedirs_p()
+    (user_plugin_basepath / "__init__.py").write_text(INIT_PY)
+    (user_plugin_path / "__init__.py").write_text(INIT_PY)

```

Comment:
What happens with this code if this structure already exists inside `~/.spyder`?